pipeline/relation_embedder/relation_embedder/bash_secrets_extension.sh

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env bash
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+set -euo pipefail
+
+OWN_FILENAME="$(basename $0)"
+LAMBDA_EXTENSION_NAME="$OWN_FILENAME" # (external) extension name has to match the filename
+TMP_FILE=/tmp/$OWN_FILENAME
+
+# Graceful Shutdown
+_term() {
+  echo "[${LAMBDA_EXTENSION_NAME}] Received SIGTERM"
+  # forward SIGTERM to child procs and exit
+  kill -TERM "$PID" 2>/dev/null
+  echo "[${LAMBDA_EXTENSION_NAME}] Exiting"
+  exit 0
+}
+
+forward_sigterm_and_wait() {
+  trap _term SIGTERM
+  wait "$PID"
+  trap - SIGTERM
+}
+
+# Initialization
+# To run any extension processes that need to start before the runtime initializes, run them before the /register
+echo "[${LAMBDA_EXTENSION_NAME}] Initialization"
+
+# Read secret from AWS Secrets Manager
+get_secret() {
+  local secret_name=$1
+  local secret_value
+  echo "[${LAMBDA_EXTENSION_NAME}] Getting secret: $secret_name" > /dev/tty
+
+  secret_value=$(aws secretsmanager get-secret-value --secret-id "$secret_name" --query SecretString --output text)
+  echo "$secret_value"
+}
+
+# Extract secret value from environment variable and write to config file
+extract_secret_value_to_config() {
+  local env_var=$1
+  local env_var_value
+  local env_var_key
+  local secret_key
+  local secret_value
+
+  env_var_value=$(echo "$env_var" | cut -d= -f2)
+  env_var_key=$(echo "$env_var" | cut -d= -f1)
+
+  if [[ $env_var_value == "secret:"* ]]; then
+    secret_key=$(echo "$env_var_value" | cut -d: -f2)
+    secret_value=$(get_secret "$secret_key")
+    echo "$env_var_key=\"$secret_value\"" >> /tmp/config
+  fi
+}
+
+# Create a configuration file with secrets
+create_config_file() {
+  local env_vars
+  local env_var
+
+  # Ensure the file exists, and is empty
+  touch /tmp/config
+  echo -n > /tmp/config
+
+  env_vars=$(printenv)
+  for env_var in $env_vars; do
+    extract_secret_value_to_config "$env_var"
+  done
+}
+
+create_config_file
+
+# Registration
+# The extension registration also signals to Lambda to start initializing the runtime.
+HEADERS="$(mktemp)"
+echo "[${LAMBDA_EXTENSION_NAME}] Registering at http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register"
+curl -sS -LD "$HEADERS" \
+  -XPOST "http://${AWS_LAMBDA_RUNTIME_API}/2020-01-01/extension/register" \
+  --header "Lambda-Extension-Name: ${LAMBDA_EXTENSION_NAME}" \
+  -d "{ \"events\": [\"SHUTDOWN\"]}" > "$TMP_FILE"
+
+RESPONSE=$(<"$TMP_FILE")
+# Extract Extension ID from response headers
+EXTENSION_ID=$(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)
+echo "[${LAMBDA_EXTENSION_NAME}] Registration response: ${RESPONSE} with EXTENSION_ID $(grep -Fi Lambda-Extension-Identifier "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)"
+
+# Event processing
```

Comment:
This event loop seems to be necessary to keep the Lambda running happily. The extension is invoked as a separate process from the main lambda, but the extension API will fail the start-up if we exit before the lifetime of the lambda container itself.

There may be scope to simplify this!

Reply:
Would sleeping for 15 minutes work here, or does the extension have to explicitly listen for the shutdown event to avoid keeping the Lambda awake?

Reply:
After some experiments, removing the `while` loop here results in the following error at lambda initialization: 
<img width="1660" alt="Screenshot 2024-12-12 at 15 23 22" src="https://github.com/user-attachments/assets/fee4a50a-7685-4b0a-9491-bdccb35039d0" />

It looks like exiting early results in stopping the runtime interface emulator (RIE) from invoking the main lambda.

> Would sleeping for 15 minutes work here

Putting a `sleep 900` instead of the loop, does allow the lambda to be invoked and terminated (though the logs do mutter about a forced stop). However a 2nd invocation then fails with the following error:

<img width="1664" alt="Screenshot 2024-12-12 at 15 29 14" src="https://github.com/user-attachments/assets/8d21fdec-999d-4a1e-9aab-b2b236395d24" />

So I think that messes with some logic around invocation in the RIE in a way that may be reproduced in an in-situ invocation.

Reply:
That makes sense. I expect it starts when a lambda is first switched on and keeps going until there's nothing left to do, rather than running for each invocation.  This allows warm starts to benefit the most from that warmth.